### PR TITLE
[BEAM-10769] Clarify Avro IO documentation. 

### PR DESCRIPTION
**Pull request by tvalentyn**
Our Avro IO documentation instructs the user to pass the schema parsed by avro.schema.Parse. On Python 3, we use FastAvro module, which does not accept avro-python3 schema classes. This PR corrects this instruction. 

To make Py2->3 transition easier for the users, we try to catch if users continue to feed avro-python3 schemas into Avro IO and error-out with a better message.

Also, fastavro does not require a parsed schema, since the schema is re-parsed anyway, see: [1], [2], it would be simpler to ask users to pass raw schema.

[1] https://github.com/apache/beam/blob/1b6b60e9a690784ead18881d48a8dc3ab2ea95ad/sdks/python/apache_beam/io/avroio.py#L634
[2] https://github.com/fastavro/fastavro/blob/9b0cb9154df13e7bcf1b3eaf88c8ae7d6065666a/fastavro/_write_py.py#L381

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels
Python tests

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tvalentyn** [comment]
Run Python 3.6 postcommit

**Comment by tvalentyn** [comment]
Run Python 3.6 postcommit

**Comment by tvalentyn** [comment]
R: @chamikaramj

**Comment by tvalentyn** [comment]
Run Python 3.6 postcommit

**Comment by tvalentyn** [comment]
Run Python 3.6 PostCommit

**Comment by tvalentyn** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/io/avroio.py [hunk 1]
```diff
@@ -90,7 +78,7 @@ def __init__(
       file_pattern=None,
       min_bundle_size=0,
       validate=True,
-      use_fastavro=_use_fastavro()):
+      use_fastavro=True):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/avroio.py:81]: So, to clarify, does both Avro and fastavro work for Python even though we set fastavro to default ?

File: sdks/python/apache_beam/io/avroio.py [hunk 2]
```diff
@@ -511,7 +488,9 @@ def __init__(
         end in a common extension, if given by file_name_suffix. In most cases,
         only this argument is specified and num_shards, shard_name_template, and
         file_name_suffix use default values.
-      schema: The schema to use, as returned by avro.schema.Parse
+      schema: The schema to use (dict).
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/avroio.py:491]: Does fastavro support a more structured definition for schemas or does is have to be a dict ?

**Review by chamikaramj — COMMENTED** [review]
Thanks!

Sorry about the delay.

File: sdks/python/apache_beam/io/avroio.py (hunk 2 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py:491]: According to https://fastavro.readthedocs.io/en/latest/schema.html, it's a dict.

**Comment by tvalentyn** [comment]
Will merge this after we remove Py2 support.

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py:81]: It's possible to force FastAvro -> Avro, although we currently have avro==1.8.2 on Dataflow workers, which is outdated and may have Py3 bugs. It's unclear if it's safe to update.
There is a comment: 

    # Do not upgrade avro library to 1.9.0 until
    # https://issues.apache.org/jira/browse/AVRO-2429 has been
    # resolved (see http://b/133317612 for more context)

Comment (chamikaramj) [reply sdks/python/apache_beam/io/avroio.py:81]: This was fixed in Avro 1.9.2: https://issues.apache.org/jira/browse/AVRO-2429
So it should be safe to upgrade to that version.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Comment by tvalentyn** [comment]
retest this please

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Pull request merged**